File: frontend/src/types/market.ts
export interface L2Level {
  price: number
  qty: number
  orders: number
}

export interface Snapshot {
  security_id: number
  symbol: string
  ts: string
  best_bid: number | null
  best_ask: number | null
  bid_qty: number | null
  ask_qty: number | null
  spread: number | null
  mid: number | null
  last_trade_price: number | null
  last_trade_qty: number | null
  vwap: number | null
  total_traded_qty: number | null
  total_traded_value: number | null
  bids: L2Level[]
  asks: L2Level[]
}

export interface Trade {
  security_id: number
  symbol: string
  ts: string
  price: number
  qty: number
  aggressor_side: 'B' | 'S'
}

export interface Signal {
  symbol: string
  signal: string
  direction: 'BUY' | 'SELL' | 'NEUTRAL'
  strength: number
  ts: string
  detail: Record<string, unknown>
}

export interface WsMessage {
  type: 'snapshot' | 'trade'
  security_id: number
  symbol: string
  ts: number
  [key: string]: unknown
}

// ── Static reference data for dashboard panels ────────────────────────────────

export interface FlowEntry {
  label: string
  value: number      // R$ millions
  change: number
}

export interface IndexEntry {
  label: string
  value: number
  change: number
  max: number
  min: number
}

export interface DIFuture {
  code: string       // e.g. "DI1F26"
  rate: number
  change: number
}

export interface WorldMarket {
  symbol: string
  value: number
  change: number
}

export interface Commodity {
  symbol: string
  value: number
  change: number
  unit: string
}

export interface StockRow {
  symbol: string
  last: number
  change: number
  volume: number
}

File: backend/app/schemas/market.py
from datetime import datetime
from pydantic import BaseModel


class L2LevelOut(BaseModel):
    price: float
    qty: int
    orders: int


class SnapshotOut(BaseModel):
    security_id: int
    symbol: str
    ts: datetime
    best_bid: float | None
    best_ask: float | None
    bid_qty: int | None
    ask_qty: int | None
    spread: float | None
    mid: float | None
    last_trade_price: float | None
    last_trade_qty: int | None
    vwap: float | None
    total_traded_qty: int | None
    total_traded_value: float | None
    bids: list[L2LevelOut] = []
    asks: list[L2LevelOut] = []


class TradeOut(BaseModel):
    security_id: int
    symbol: str
    ts: datetime
    price: float
    qty: int
    aggressor_side: str


class InstrumentOut(BaseModel):
    security_id: int
    symbol: str
    security_type: str
    currency: str
    lot_size: int
    maturity_date: str | None = None
    strike_price: float | None = None
    put_or_call: str | None = None
